prog/prog.go

```diff
@@ -212,45 +254,54 @@ func (p *Prog) insertBefore(c *Call, calls []*Call) {
 }
 
 // replaceArg replaces arg with arg1 in call c in program p, and inserts calls before arg call.
-func (p *Prog) replaceArg(c *Call, arg, arg1 *Arg, calls []*Call) {
-	if arg.Kind != ArgConst && arg.Kind != ArgResult && arg.Kind != ArgPointer && arg.Kind != ArgUnion {
-		panic(fmt.Sprintf("replaceArg: bad arg kind %v", arg.Kind))
-	}
-	if arg1.Kind != ArgConst && arg1.Kind != ArgResult && arg1.Kind != ArgPointer && arg.Kind != ArgUnion {
-		panic(fmt.Sprintf("replaceArg: bad arg1 kind %v", arg1.Kind))
-	}
-	if arg.Kind == ArgResult {
-		delete(arg.Res.Uses, arg)
-	}
+func (p *Prog) replaceArg(c *Call, arg, arg1 Arg, calls []*Call) {
 	for _, c := range calls {
 		sanitizeCall(c)
 	}
 	p.insertBefore(c, calls)
 	// Somewhat hacky, but safe and preserves references to arg.
-	uses := arg.Uses
-	*arg = *arg1
-	arg.Uses = uses
-	if arg.Kind == ArgResult {
-		delete(arg.Res.Uses, arg1)
-		arg.Res.Uses[arg] = true
+	uses := *arg.Uses()
+	switch a := arg.(type) {
+	case *ConstArg:
+		a1 := arg1.(*ConstArg)
+		*a = *a1
+	case *ResultArg:
+		a1 := arg1.(*ResultArg)
+		if a.Res != nil {
+			delete(*a.Res.Uses(), arg)
+		}
+		*a = *a1
+	case *PointerArg:
+		a1 := arg1.(*PointerArg)
+		*a = *a1
+	case *UnionArg:
+		a1 := arg1.(*UnionArg)
+		*a = *a1
+	default:
+		panic(fmt.Sprintf("replaceArg: bad arg kind %v", arg))
+	}
+	*arg.Uses() = uses
+	if a, ok := arg.(*ResultArg); ok && a.Res != nil {
```

Comment:
Is it possible to fold this into the switch above? We already did type switch and know when it's ResultArg.
Will probably also help to get rid of uses for all arg types.